### PR TITLE
config: add C99 feature tests

### PR DESCRIPTION
**Pull request by hzhou**
The idea is run `PAC_CHECK_C99`, which tests a list of required C99
features. The configure will fail if one of these features are not
supported.

TO DO:  Test against a list of compilers that we want to support.
        Add feature tests that we want to use/require.
        Comment out feature tests that we want to relax.

### CODE REVIEW

**Comment by hzhou** [comment]
Tested on Mac running clang (as gcc):
```
checking for icc... no
checking for pgcc... no
checking for xlc... no
checking for xlC... no
checking for pathcc... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
CC: gcc, CFLAGS: -O2
checking C99 inline feature... yes
checking C99 in-code variable declaration feature... yes
checking C99 variable declaration in for-loop feature... yes
checking C99 inline comment feature... yes
checking C99 bool feature... yes
checking C99 stdint feature... yes
checking C99 inttypes feature... 00000000000000000042 yes
checking C99 variadic macro feature... yes
checking C99 compound literal feature... yes
checking C99 designated init feature... yes
checking C99 __func__ feature... yes
configure: creating ./config.status
```

File: confdb/aclocal_c99.m4 [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    PAC99_VARIADIC_MACRO
+    PAC99_COMPOUND_LITERAL
+    PAC99_DESIGNATED_INIT
+    PAC99__FUNC__
+])
+
+dnl ***************************
+AC_DEFUN([PAC99_TEST], [
+    AC_MSG_CHECKING([C99 $1 feature])
+    AC_TRY_RUN([$2],
```
Comment (raffenet) [inline confdb/aclocal_c99.m4]: Are compile or link tests sufficient for all these features? Also, the `AC_TRY_` macros are deprecated. You should use `AC_COMPILE|LINK_IFELSE`. https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/html_node/Obsolete-Macros.htm

Comment (hzhou) [reply confdb/aclocal_c99.m4]: I was thinking on wether to use `RUN` or `LINK`. Certain feature, e.g. `snprintf` is important to test its return value; but `Run` definitely won't work for cross compiling. Since I have separate test scripts that test individual compilers, I guess in configure, just link should be sufficient.

I'll update it with `AC_LINK_IFELSE`.

Comment (pavanbalaji) [reply confdb/aclocal_c99.m4]: `TRY_RUN` is not cross-compile safe.

Comment (raffenet) [reply confdb/aclocal_c99.m4]: I believe it is also best-practice to use `AC_LANG_PROGRAM` to create the tests.

Comment (hzhou) [reply confdb/aclocal_c99.m4]: `AC_LANG_PROGRAM` prepends the code with `#defines` configured so far. Here we are testing raw c99 features, I don't think we need bother with it.

Comment (hzhou) [reply confdb/aclocal_c99.m4]: Just found out there is `AC_PROG_CC_C99` that tests most of these features in one go and add necessary option as well. So we only need test a minimum set when `AC_PROG_CC_C99` returns `no`.

**Comment by hzhou** [comment]
Merged #3754
